### PR TITLE
feat(Rust): Use rename_all

### PR DESCRIPTION
**Pull request by taorepoara**
Detects the the most used naming style in the class properties and enum cases to define it in the rename_all when needed.

Replaces https://github.com/quicktype/quicktype/pull/2210
Closes https://github.com/quicktype/quicktype/issues/528

### CODE REVIEW

**Comment by taorepoara** [comment]
I don't understand the test error. Can someone help me with this ?

**Comment by taorepoara** [comment]
Ok, I understood the problem.

The `union` field of this schema test is renamed to `top_level_union`: https://github.com/quicktype/quicktype/blob/master/test/inputs/schema/pattern.schema#L12

Does someone know why the field name is not kept ?

**Comment by taorepoara** [comment]
I could fix the problem with the unit tests.

I also add one with camel case properties to test the new feature.

File: packages/quicktype-core/src/language/Rust.ts [hunk 1]
```diff
@@ -314,11 +374,23 @@ export class RustRenderer extends ConvenienceRenderer {
...
+        // List the possible naming styles for every class property
+        const propertiesNamingStyles: { [key: string]: string[] } = {};
+        this.forEachClassProperty(c, "none", (_name, jsonName, _prop) => {
+            propertiesNamingStyles[jsonName] = listMatchingNamingStyles(jsonName);
+        });
+
+        // Set the default naming style on the struct
+        const defaultStyle = "snake_case";
+        const preferedNamingStyle = getPreferedNamingStyle(Object.values(propertiesNamingStyles).flat(), defaultStyle);
+        if (preferedNamingStyle !== defaultStyle)
```
Comment (dvdsgl) [inline packages/quicktype-core/src/language/Rust.ts:386]: Please add braces around if body.

File: packages/quicktype-core/src/language/Rust.ts [hunk 2]
```diff
@@ -52,6 +52,64 @@ export const rustOptions = {
...
+    },
+    "lowercase": {
+        regex: /^[a-z][a-z0-9]*$/,
+        toParts: (name: string): string[] => [name],
+        fromParts: (parts: string[]): string => parts.map(p => p.toLowerCase()).join(''),
+    },
+    "UPPERCASE": {
+        regex: /^[A-Z][A-Z0-9]*$/,
+        toParts: (name: string): string[] => [name],
+        fromParts: (parts: string[]): string => parts.map(p => p.toUpperCase()).join(''),
```
Comment (dvdsgl) [inline packages/quicktype-core/src/language/Rust.ts:109]: You don't need to write the types for each toParts and fromParts wince you declared them for the entire struct.

File: packages/quicktype-core/src/language/Rust.ts [hunk 3]
```diff
@@ -52,6 +52,64 @@ export const rustOptions = {
     leadingComments: new BooleanOption("leading-comments", "Leading Comments", true)
 };
 
+type nameToParts = (name: string) => string[];
```
Comment (dvdsgl) [inline packages/quicktype-core/src/language/Rust.ts:55]: Type names should be like NameToParts.

File: packages/quicktype-core/src/language/Rust.ts [hunk 4]
```diff
@@ -52,6 +52,64 @@ export const rustOptions = {
...
+type partsToName = (parts: string[]) => string;
+type namingStyle = {
+    regex: RegExp,
+    toParts: nameToParts,
+    fromParts: partsToName,
+};
+
+const NamingStyle: { [key: string]: namingStyle } = {
```
Comment (dvdsgl) [inline packages/quicktype-core/src/language/Rust.ts:63]: Values should be named like namingStyle.

Comment (dvdsgl) [reply packages/quicktype-core/src/language/Rust.ts:63]: The type should be Record<string, NamingStyle>

**Review by dvdsgl — CHANGES_REQUESTED** [review]
Very nice work! Just a few small stylistic changes requested.

File: packages/quicktype-core/src/language/Rust.ts (hunk 2 above)
Comment (taorepoara) [reply packages/quicktype-core/src/language/Rust.ts:109]: Actually I use them to detect the input fields formats.
I also use them to detect if a field does not match the global type.

**Comment by dvdsgl** [comment]
I’ve just fixed the tests, please merge master.

**Pull request merged**